State Migration (#454)
klaytn/klaytn-docs: docs/klaytn/design/storage.md

# Storage Layer <a id="storage-layer"></a>

## State Migration <a id="block-archiving"></a>

> As more blocks are added to the blockchain, chain data also pile up. Chain data are necessary for node operation, so  they are stored in the node storage as a data structure called trie, and ultimately in a database called LevelDB. So with more blocks, comes more chain data in the storage, along with increasing cost. Klaytn therefore provides a feature called State Migration that allows you to reduce the amount of required storage space.

Reviewer: ```suggestion
As more blocks are added to the blockchain, chain data also pile up. Chain data are necessary for node operation, so they are stored in the node storage as a data structure called trie, and ultimately in a database called LevelDB. So with more blocks, comes more chain data in the storage, along with increasing cost. Klaytn, therefore, provides a feature called State Migration that allows you to reduce the amount of required storage space.
```

> State Migration targets state tries, which comprise most of the chain data. It deletes state trie nodes that are not required for processing new blocks. It only leaves the state trie nodes starting from a specific block. After State Migration, you are only left with the latest data required for EN operation, consisting of state tries of the target block as well as newly added blocks.

Reviewer: There is a limitation to this feature. How about adding a warning like this 
```
Note that a node can read an old state which is earlier than the migration starting bock after State Migration finished. In other words, the node can't return a balance at the old block number through the `klay_getBalance` API.
```

# Storage Layer <a id="storage-layer"></a>

## State Migration <a id="block-archiving"></a>

As more blocks are added to the blockchain, chain data also pile up. Chain data are necessary for node operation, so they are stored in the node storage as a data structure called trie, and ultimately in a database called LevelDB. So with more blocks, comes more chain data in the storage, along with increasing cost. Klaytn, therefore, provides a feature called State Migration that allows you to reduce the amount of required storage space.

> State Migration targets state tries, which comprise most of the chain data. It deletes state trie nodes that are not required for processing new blocks. It only leaves the state trie nodes starting from a specific block. After State Migration, you are only left with the latest data required for node synchronization, consisting of state tries of the target block as well as newly added blocks.

Reviewer 2: ```suggestion
State Migration targets state tries, which comprise most of the chain data. It deletes state trie nodes that are not required for processing new blocks. It only leaves the state trie nodes reachable from the state trie root of a specific block. After State Migration, you are only left with the latest data required for node synchronization, consisting of state trie nodes of the target block as well as newly added blocks.
```

> State Migration targets state tries, which comprise most of the chain data. It deletes state trie nodes that are not required for processing new blocks. It only leaves the state trie nodes �reachable from the state trie root of a specific block. After State Migration, you are only left with the latest data required for node synchronization, consisting of state trie nodes of the target block as well as newly added blocks.

Reviewer 2: ```suggestion
State Migration targets state tries, which comprise most of the chain data. It deletes state trie nodes that are not required for processing new blocks. It only leaves the state trie nodes reachable from the state trie root of a specific block. After State Migration, you are only left with the latest data required for node synchronization, consisting of state trie nodes of the target block as well as newly added blocks.
```

Note that a node can't read old states from blocks previous to the target block after State Migration. In other words, you can't return the balance from an old block number using the `klay_getBalance` API.

More details on the mechanism of State Migration can be found below:
[Klaytn v1.5.0 State Migration: Saving Node Storage](https://medium.com/klaytn/klaytn-v1-5-0-state-migration-saving-node-storage-1358d87e4a7a)
[Klaytn State Migration: An Efficient Way to Reduce Blockchain Data](https://medium.com/klaytn/klaytn-state-migration-an-efficient-way-to-reduce-blockchain-data-6615a3b36523)

> To use State Migration, please refer to the documentation on admin APIs.[`startStateMigration`](https://docs.klaytn.foundation/dapp/json-rpc/api-references/admin#admin_startstatemigration) lets you start State Migration, [`stateMigrationStatus`](https://docs.klaytn.foundation/dapp/json-rpc/api-references/admin#admin_statemigrationstatus) allows you to check the Migration status, and [`stopStateMigration`](https://docs.klaytn.foundation/dapp/json-rpc/api-references/admin#admin_stopstatemigration) allows you to stop the process.

Reviewer 2: ```suggestion
To use State Migration, please refer to the documentation on admin APIs.[`admin_startStateMigration`](https://docs.klaytn.foundation/dapp/json-rpc/api-references/admin#admin_startstatemigration) lets you start State Migration, [`admin_stateMigrationStatus`](https://docs.klaytn.foundation/dapp/json-rpc/api-references/admin#admin_statemigrationstatus) allows you to check the Migration status, and [`admin_stopStateMigration`](https://docs.klaytn.foundation/dapp/json-rpc/api-references/admin#admin_stopstatemigration) allows you to stop the process.
```
